src/date/date.rs

```diff
@@ -0,0 +1,283 @@
+#[macro_use]
+extern crate clap;
+extern crate uucore;
+
+use chrono::{DateTime, FixedOffset, Offset, Local};
+use chrono::offset::utc::UTC;
+use std::fs::File;
+use std::io::{BufRead, BufReader};
+use std::path::PathBuf;
+
+// Options
+const DATE: &'static str = "date";
+const HOURS: &'static str = "hours";
+const MINUTES: &'static str = "minutes";
+const SECONDS: &'static str = "seconds";
+const NS: &'static str = "ns";
+
+// Help strings
+
+static ISO_8601_HELP_STRING: &'static str = "output date/time in ISO 8601 format.
+ FMT='date' for date only (the default),
+ 'hours', 'minutes', 'seconds', or 'ns'
+ for date and time to the indicated precision.
+ Example: 2006-08-14T02:34:56-06:00";
+
+static RFC_2822_HELP_STRING: &'static str = "output date and time in RFC 2822 format.
+ Example: Mon, 14 Aug 2006 02:34:56 -0600";
+
+static RFC_3339_HELP_STRING: &'static str = "output date/time in RFC 3339 format.
+ FMT='date', 'seconds', or 'ns'
+ for date and time to the indicated precision.
+ Example: 2006-08-14 02:34:56-06:00";
+
+/// Settings for this program, parsed from the command line
+struct Settings {
+    utc: bool,
+    format: Format,
+    date_source: DateSource,
+    set_to: Option<DateTime<FixedOffset>>,
+}
+
+/// Various ways of displaying the date
+enum Format {
+    Iso8601(Iso8601Format),
+    Rfc2822,
+    Rfc3339(Rfc3339Format),
+    Custom(String),
+    Default,
+}
+
+/// Various places that dates can come from
+enum DateSource {
+    Now,
+    Custom(String),
+    File(PathBuf),
+}
+
+enum Iso8601Format {
+    Date,
+    Hours,
+    Minutes,
+    Seconds,
+    Ns,
+}
+
+impl<'a> From<&'a str> for Iso8601Format {
+    fn from(s: &str) -> Self {
+        match s {
+            HOURS => Iso8601Format::Hours,
+            MINUTES => Iso8601Format::Minutes,
+            SECONDS => Iso8601Format::Seconds,
+            NS => Iso8601Format::Ns,
+            DATE => Iso8601Format::Date,
+            // Should be caught by clap
+            _ => panic!("Invalid format: {}", s),
+        }
+    }
+}
+
+enum Rfc3339Format {
+    Date,
+    Seconds,
+    Ns,
+}
+
+impl<'a> From<&'a str> for Rfc3339Format {
+    fn from(s: &str) -> Self {
+        match s {
+            DATE => Rfc3339Format::Date,
+            SECONDS => Rfc3339Format::Seconds,
+            NS => Rfc3339Format::Ns,
+            // Should be caught by clap
+            _ => panic!("Invalid format: {}", s),
+        }
+    }
+}
+
+pub fn uumain(args: Vec<String>) -> i32 {
+
+    let settings = parse_cli(args);
+
+    if let Some(_time) = settings.set_to {
+        unimplemented!();
+        // Probably need to use this syscall:
+        // https://doc.rust-lang.org/libc/i686-unknown-linux-gnu/libc/fn.clock_settime.html
+
+    } else {
+        // Declare a file here because it needs to outlive the `dates` iterator.
+        let file: File;
+
+        // Get the current time, either in the local time zone or UTC.
+        let now: DateTime<FixedOffset> = match settings.utc {
```

Comment:
Probably because I was having too much fun with match expressions. You're right; an if is cleaner.